crates/chain/src/keychain/txout_index.rs

```
@@ -548,114 +444,146 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
         }
     }
 
-    fn replenish_lookahead(&mut self, keychain: &K, lookahead: u32) {
-        let descriptor_opt = self.keychains_to_descriptors.get(keychain).cloned();
-        if let Some((descriptor_id, descriptor)) = descriptor_opt {
-            let next_store_index = self.next_store_index(descriptor_id);
-            let next_reveal_index = self.last_revealed.get(&descriptor_id).map_or(0, |v| *v + 1);
+    fn replenish_lookahead_did(&mut self, did: DescriptorId) {
+        if let Some(keychain) = self.descriptor_ids_to_keychains.get(&did).cloned() {
+            self.replenish_lookahead(&keychain, self.lookahead);
+        }
+    }
```

I don't think this method is necessary. We are only using it in 2 places; `Indexer::index_txout` and `apply_changeset`.

For `index_txout`, the keychain is already available for us (however we may need to clone the `inner.scan_txout` result because of the referencing rules of rust).

For `apply_changeset`, we can just query `Self::descriptor_ids_to_keychains` within the loop.

---

I think the method is useful to explicitly make the choice to silently ignore the corresponding `did` or keychain not existing. Doing the lookups in place means me or someone else is likely to just do `.expect("invariant")` in that case without understanding context and reviewers are unlikely to notice it.

I take your point that in one place the keychain already exists. I refactored the methods to take advantage of that.